app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```
@@ -0,0 +1,239 @@
+        for (int i=0; i<tokens.length; i++) {
+            String token = tokens[i];
+
+            // Every 2nd token is a attribute enclosed in {{ }}
+            if (i % 2 == 1) {
+
+                // Split token into tokenSplits bounded by < and >
+                String[] tokenSplits = TemplatePrinterUtils.splitKeepDelimiter(token, "<|(\\}{2})", ">|(\\{{2})");
+
+                // First and last tokenSplits are {{ and }}
+                for (int j=1; j<tokenSplits.length-1; j++) {
+                    String tokenSplit = tokenSplits[j];
+
+                    // tokenSplit is key or whitespace
+                    if (!tokenSplit.startsWith("<")) {
+
+                        // Remove whitespace from key
+                        String key = TemplatePrinterUtils.remove(tokenSplit, " ");
+
+                        if (values.containsKey(key) && (key = values.getString(key)) != null) {
+                            // Populate with value
+                            tokenSplits[j] = key;
+                        } else {
+                            // Empty if not found
+                            tokenSplits[j] = "";
+                        }
+                    }
+
+                }
+
+                // Remove {{ and }}
+                tokenSplits[0] = "";
+                tokenSplits[tokenSplits.length-1] = "";
+
+                // Reconstruct token
+                tokens[i] = TemplatePrinterUtils.join(tokenSplits);
+            }
+        }
+
+        // Reconstruct input
+        return TemplatePrinterUtils.join(tokens);
+    }
+
+    /**
+     * Validates the input string for well-formed {{ }} and < > pairs.
+     * If malformed, throws a RuntimeException which will be caught by
+     * doInBackground() and passed to the attached PopulateListener.onError()
+     * 
+     * @param input String to validate
+     */
+    private static void validateStringOrThrowException(String input) {
```

Yeah, not entirely sure why I used a boolean flag and a break statement. I would suggest at every place where there's currently a "isWellFormed = false; break;", you could throw an exception with details about placement and type of pattern.